Telemetry fix for eof

aaithal opened the pull request:
… ack metric is received; This fixes the error of reading from a closed connection and getting the "connection closed" error message. The client correctly returns io.EOF when remote server closes connection.

r? @samuelkarp @euank

euank commented on agent/tcs/handler/handler.go:
```diff
@@ -95,8 +95,17 @@ func startSession(url string, region string, credentialProvider credentials.AWSC
+	// start a timer and listens for tcs heartbeats/acks. The timer is reset when
+	// we receive a heartbeat from the server or when a publish metrics message
+	// is acked.
+	timer := time.AfterFunc(utils.AddJitter(heartbeatTimeout, heartbeatJitter), func() {
```
I think you want to throw in a defer timer.Stop to prevent logging the below even after returning from this function (which will only happen if the client is closed anyways).

I think the acs client has this problem as well.

euank commented on agent/wsclient/client.go:
```diff
@@ -205,6 +205,7 @@ func (cs *ClientServerImpl) ConsumeMessages() error {
 		messageType, message, cerr := cs.Conn.ReadMessage()
 		err = cerr
 		if err != nil {
+			log.Error("Error in ConsumeMessages", "messageType", messageType, "err", err.Error(), "message", message)
```
It was intentional that if `err == io.EOF` nothing was logged at the `Error` level. It used to log at that level which resulted in noisy 'err' logs that were simply a normal server close.

euank commented:
Couple comments around logging cleanliness which I think should be addressed. Other than those, looks good... Though I wouldn't mind a quick explanation of how/why this fixes the issue since just reading the diff doesn't make the root cause here clear to me.

aaithal commented:
Addressed comments. The root cause of the issue is that TCS won't send heatbeats to agent as heartbeats are sent by the server when it detects an idle connection. Since the agent is constantly sending metric data to server, agent would never receive heartbeats, which would lead to timer closing the connection and then websocket client writing/reading from a closed connection, which results in net. errClosing as the client was the entity that closed the connection. io.EOF means server closing the connection, which is being handled properly in the code.

euank commented:
Ah, thanks for explaining @aaithal.
I was confused in thinking the default return change in `Close` was related.

LGTM

samuelkarp commented:
:+1:
